museekd/museekd.1
.TH "MUSEEKD" "1" "Release 0.2.0" "daelstorm" "Museek Daemon Plus"
.SH "NAME"
.LP 
\fBMuseekd\fR \- C++ Daemon/Client for Soulseek
.SH "SYNOPSIS"
.B museek
[\-c <\fIfilename\fP>] [\-\-config <\fIfilename\fP>]
[\-V] [\-\-version]
[\-h] [\-\-help]
.SH "DESCRIPTION"
.LP 
Museekd is a headless client (daemon) for the Soulseek P2P network and for Soulfind Servers. It is configured with musetup and can be controlled with variety of interfaces and GUIs, such as Museeq and Mucous. It handles any chatting, transferring files, searching, managing of users or browsing of shares requested by the various interfaces or GUIs.
.LP 
Before running, you will need a configure it with \fImusetup\fP(1).
.SH "OPTIONS"
.LP 
Museekd accepts the following options:
.TP 
.B \-c <\fIfilename\fP>, \-\-config <\fIfilename\fP>
Use a different config file.
.TP 
.B \-V, \-\-version
Display Version and exit.
.TP 
.B \-d, \-\-debug
Show debugging messages.
.TP 
.B \-h, \-\-help
Display Help and exit.

.SH "EXAMPLE"
.LP 
To run this program the standard way type:
.LP 
museekd
.LP 
Alternativly you use a different config file with:
.LP 
museekd \-\-config config.xml
.SH "FILES"
.TP 
 \fI~/.museekd/config.xml\fR
The default location for the \fBmuseekd\fP config file.
.SH "AUTHORS"
.LP 
Hyriand <hyriand@thegraveyard.org>
.LP 
daelstorm <daelstorm@gmail.com>
.SH "SEE ALSO"
.LP 
\fImucous\fP(1) \fImulog\fP(1) \fImurmur\fP(1) \fImuscan\fP(1) \fImuscand\fP(1) \fImuseekcontrol\fP(1) \fImuseeq\fP(1) \fImusetup\fP(1) \fImusetup\-gtk\fP(1)


CHANGELOG
Museek Plus Changelog

0.1.13:  August 25th 2007

BUILD SYSTEMS
CMake, qmake, SCons and Python's distutils
 * RECOMMENDED: cmake
 * CMake compiles and installs everything. Read the INSTALL files for details.
 * qmake compiles just museeq. Read museeq/INSTALL
 * You can install the python bindings, apps, and clients with distutils.
   Run "python setup.py install" in each python sub-directory
 * SCons is still available, but not recommended. Read INSTALL.scons

MUSEEK DAEMON
 * Ignore SIGPIPE (to prevent crashes on BSDs)
 * Catch std::bad_alloc when unpacking a network message.
 * Always get IP from server, when requested
 * Autoretry downloads
 * Syslog support
 * Watch trusted users' status

MUSEEQ
 * Settings revamped
 * Shares configuration added to Settings
 * Added Audacioous to NowPlaying QSA script
 * Added colors to transfers' status column

MUCOUS
 * Setup revamped
 * Many, many bugfixes

MUSEEKCONTROL
 * Implemented a few more unfinished options

MURMUR
 * The basic features are here, enough for inclusion in this release.

0.1.12:  October 21th 2006

MUSEEK DAEMON
 * Fixed museekd not reacting on connect mode changes during run-time (hyriand)
 * Fixed the clear-initiating-transfer that causes a delayed crash (hyriand)
 * Character unpacking checks (fixes corrupted message crashes) (hyriand)
 * Trusted Users' status is now watched
 * IMPORTANT NOTE: the CFLAGS variable renamed to FLAGS and is now a String 
   instead of List, your mulocal.py will need to be modified or erased to allow
   SCons to run correctly. Not doing so will result in SCons failing at the 
   first compile-test.
 * Change the way the IP Address string is created to fix endian bug (hyriand) 
 * New SCons variable LOCKFLAGS disables SCons' automatic flag adding
 * Replaced depreciated call to Send Speed (34) on download completion
   with Send Upload Speed (121) on upload completion. (daelstorm)
 * Check for read/write permissions to config file before loading (daelstorm)
 * Added Global message (66) handler, and pass it to the clients (daelstorm)
 * Removed disconnect-on-hangup-signal (caused a loop)
 * Client message 'Download file to a directory' added (daelstorm)
 * Sends a Museek Daemon version message to clients on connection (daelstorm)
 * Sends the number of privileged users to clients on Server Login (daelstorm)

MUSCAN
 * Ignore .dotfiles and .dotdirs/
 * Removing trailing /s for shared directories
 
MUSEEQ
 * Fixed crash when disconnecting while still connecting to museekd by using
   "mSocket->deleteLater()" instead of "delete mSocket" (hyriand)
 * Fixed method of creating UNIX socket (hyriand)
 * Check option in settings to toggle display of exit dialog
 * Added nick's transfer list color patch transfers are now different colors,
   depending upon their state.
 * Added MUSEEQTRAYICON scons option to enable/disable trayicon at buildtime.
 * Moved Trayicon into Museeq class, made it change with away status and added
   menu option to toggle it.
 * Added /commands to private chat
 * Autoconnect to daemon option added
 * IP addresses can now be viewed in the log window
 * Updated Italian translation (requiem)
 * Updated Polish translation (belegdol)
 
MUSETUP-GTK
 * Scons tries more than one method to detect PyGTK
 
MUCOUS
 * The Setup Interface is now completely controllable from the keyboard
 * Browse Shares now has expandable/collapsable directories and resizeable
   windows. There are some other UI improvements, as well.
 * Added /me and status colors to Private chat 
 * Rewrote tab completion
 * Case-insensitive sorting
 * Refactored / Renamed lots of functions
 * Killed dozens of memory leaks
 * Removed vertical borders from private chat and rooms
 * Added scrolling ticker to chat rooms
 * Improved the Transfers colors and use vertical lines instead of pipes
 * Fixed some broken config options
 * Simplified creation of setup widgets
 * Setup's control over muscan is slightly better, attempts to detect museekd
   config, to better scan correct shares databases.
 * Hundreds of minor changes
 
0.1.11:  July 7th 2006

MUSEEK DAEMON
* ReAdd the server ping, but change the delay between them from 15 to 45 seconds
  to reduce load on the server and to not get banned :).
* Try to logout when shutting down
* Fixed the minPort == maxPort bug and maxPort > minPort bug, for peer
  connections. Previously, it'd skip the port listen attempt entirely. Now it
  tries the listen on minPort (if they equal) or on maxPort (if it's smaller 
  than minport) and  then gives up if it fails. 
* SCons can now set CCFLAGS with the CFLAGS list option
* Enable or disable installation of musetup-gtk with Scons bool option:
  MUSETUPGTK
* Choose which translations to install with the MUSEEQTRANSLATIONS Scons list
  option. (New translations may require modification to the SConstruct file)

MUSEEQ
* Added Timestamp toggle to the Settings menu
* Made user status appear in Private Chat (* user is away || * user is offline )
* Ported all .ui dialogs to cpp
* Added encodings comboboxes to settings dialog, as well as making the password
  dialog protected with ****s
* Added Translation *Support*
* Le_Vert translated museeq into French
* Requiem translated museeq into Italian
* Belegdol translated museeq into Polish
* SyphiliticVagina translated museeq into Spanish
* () translated museeq into German
* Senhor_das_Moscas translated museeq into Brazilian Portuguese
* Added more icons (ip, comments, alert, and small-icons for:  browse, userinfo,
  search, chatrooms, privatechat )
* Add two of mike-labo's themes from museeq. Read the README files for credits.

MUCOUS
* Version to 0.8.2
* New commands: /autoclear, /autoretry, /beep, /trustlist
* Add trusted users list to Users mode
* Make mucous behave more nicely when not connected to a museekd.
* Reorganize parts of setup mode, add new options for buddy-only shares, etc.
* Add timestamps to Debug log, bugfixes for browse mode, topbar variables
  merged into the new storage method


museekd/museekd.cpp
/*  Museek - A SoulSeek client written in C++
    Copyright (C) 2006-2007 Ingmar K. Steen (iksteen@gmail.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif // HAVE_CONFIG_H
#include "museekd.h"
#include "configmanager.h"
#include "codesetmanager.h"
#include "servermanager.h"
#include "peermanager.h"
#include "ifacemanager.h"
#include "peersocket.h"
#include "downloadmanager.h"
#include "uploadmanager.h"
#include "sharesdatabase.h"
#include "searchmanager.h"
#include <NewNet/nnreactor.h>
#include <fstream>

Museek::Museekd::Museekd(NewNet::Reactor * reactor) : m_Reactor(reactor)
{
  /* Seed the random generator and fabricate our starting token. */
  srand(time(NULL));
  m_Token = rand();

  /* Did we get a reactor? No? Create one. */
  if(! reactor)
  {
    m_Reactor = new NewNet::Reactor();
  }

  /* Instantiate the various components. Order can be important here. */
  m_Config = new ConfigManager();
  m_Codeset = new CodesetManager(this);
  m_Server = new ServerManager(this);
  m_Peers = new PeerManager(this);
  m_Downloads = new DownloadManager(this);
  m_Uploads = new UploadManager(this);
  m_Ifaces = new IfaceManager(this);
  m_Shares = new SharesDatabase(this);
  m_BuddyShares = new SharesDatabase(this);
  m_Searches = new SearchManager(this);
}

void Museek::Museekd::LoadShares() {

  std::string shares = m_Config->get("shares", "database");
  if (! shares.empty()) {
    m_Shares->load(shares);
    m_BuddyShares->load(shares);
  }
  std::string bshares = m_Config->get("buddy.shares", "database");
  if (! bshares.empty() && haveBuddyShares())
    m_BuddyShares->load(bshares, true);
}

void Museek::Museekd::LoadDownloads() {
    m_Downloads->loadDownloads();
}

Museek::Museekd::~Museekd()
{
  /* This destructor doesn't do much. */
  NNLOG("museekd.debug", "museekd destroyed");
}

bool Museek::Museekd::isBanned(const std::string u) {
    return config()->hasKey("banned", u);
}

bool Museek::Museekd::isTrusted(const std::string u) {
    return config()->hasKey("trusted", u);
}

bool Museek::Museekd::isBuddied(const std::string u) {
    return config()->hasKey("buddies", u);
}

bool Museek::Museekd::isPrivileged(const std::string u) {
    return std::find(mPrivilegedUsers.begin(), mPrivilegedUsers.end(), u) != mPrivilegedUsers.end();
}

bool Museek::Museekd::toBuddiesOnly() {
    return config()->getBool("transfers", "only_buddies", false);
}

bool Museek::Museekd::haveBuddyShares() {
    return config()->getBool("transfers", "have_buddy_shares", false);
}

bool Museek::Museekd::trustingUploads() {
    return config()->getBool("transfers", "trusting_uploads", false);
}

bool Museek::Museekd::privilegeBuddies() {
    return config()->getBool("transfers", "privilege_buddies", false);
}

uint Museek::Museekd::upSlots() {
    return config()->getUint("transfers", "upload_slots", 0);
}

uint Museek::Museekd::downSlots() {
    return config()->getUint("transfers", "download_slots", 0);
}

// Add this user to the list of privileged ones
void Museek::Museekd::addPrivilegedUser(const std::string & user) {
    if (!isPrivileged(user)) {
        mPrivilegedUsers.push_back(user);
        NNLOG("museekd.debug", "%u privileged users", mPrivilegedUsers.size());
    }
}

// Replace the privileged users list with this new one
void Museek::Museekd::setPrivilegedUsers(const std::vector<std::string> & users) {
    mPrivilegedUsers = users;
    NNLOG("museekd.debug", "%u privileged users", mPrivilegedUsers.size());
}

void Museek::Museekd::sendSharedNumber() {
    uint32 numFiles = m_Shares->files();
    uint32 numFolders = m_Shares->folders();
    if (numFiles == 0) {
        numFiles = m_BuddyShares->files();
        numFolders = m_BuddyShares->folders();
    }

	if  (server()->loggedIn()) {
	    SSharedFoldersFiles msg(numFolders, numFiles);
	    server()->sendMessage(msg.make_network_packet());
	}
}
